### PR TITLE
Feat: use shiki syntax highlighting with custom theme

### PR DESCRIPTION
**Pull request by bholmesdev**
## What changed?
- Set syntax highlighting to Shiki
- Move our `code.css` variables to a custom Shiki theme
- Represent existing CSS variables as TypeScript tokens

## Why Shiki?

The Shikification has brought the same syntax highlighting as before, plus some added granularity! Let's break down a few of the benefits.

### Better Styles rules
Prism wraps CSS rules in a single class. This prevents highlighting for CSS-specific values like variables (ex. `--cool`) and units (ex. `rem`). Shiki offers much more granularity.

#### Before

image

#### After

image

### CSS preprocessor highlighting
Prism was tripped up by our more exotic use cases, like scss-in-astro or tailwind-directives-in-css. Shiki doesn't struggle with either of these ❤️ 

#### Before

image

image

#### After

image

image

### Escaping comments

Prism also struggled to determine the "flavor" of comments for different pieces of a given `.astro` file. Often, it would treat `<!--html comments-->` as JavaScript in need of syntax highlighting 😬 Shiki doesn't have this problem!

#### Before

Note phrases like `page` and `class` are erroneously highlighted at JS reserved words.

image

#### After

image

### Option for more features without clientside JS

My greatest complaint with Prism has been the plugin ecosystem's reliance on clientside JS. For instance, a feature [as simple as line numbers](https://prismjs.com/plugins/line-numbers/) requires a clientside bundle with zero SSR options. Since Shiki injects inline styles and disappears, we shouldn't run into these issues.

## Alternative considered: Shiki's `css-variables` theme

Shiki has one major limitation: **all theme colors must be hexcodes.** This means no `hsla` computations and no CSS variable references.

For this reason, they offer [a CSS variable theme](https://github.com/shikijs/shiki/blob/main/docs/themes.md#theming-with-css-variables) as an escape hatch. This exposes general attributes for styling via CSS in your stylesheet. This config gets us close-ish to our existing Prism config for instance:

```css
:root {
	--astro-code-color-foreground: #d8dee9;
	--astro-code-color-background: #312749;
	--astro-code-token-constant: hsla(var(--color-yellow), 1);
	--astro-code-color-text: var(--astro-code-color-foreground);
	--astro-code-token-string: hsla(var(--color-green), 1);
	--astro-code-token-comment: hsla(var(--color-gray-70), 1);
	--astro-code-token-keyword: #ff657c;
	--astro-code-token-parameter: hsla(var(--color-yellow), 1);
	--astro-code-token-function: hsla(var(--color-blue), 1);
	--astro-code-token-string-expression: hsla(var(--color-green), 1);
	--astro-code-token-punctuation: hsla(var(--color-gray-80), 1);
	--astro-code-token-link: #ff657c;
}
```

This approach was much simpler to configure from scratch. But that's because the CSS variable theme is just... far less granular. Here's a short list of limitations:
- Since there's a very finite number of variables, certain elements have to be the same color. Ex. strings and HTML tags have to be the same color, CSS properties and values have to match, JS functions and HTML attributes have to match... the list goes on.
- Prism offers a number of helpful data attributes like `language-*` for language-specific targeting. Shiki sadly lacks this luxury. So if syntax highlighting looks good in one language and needs tweaking in another, you're out of luck
- CSS variables will target a single CSS attribute like `background` or `color`. There's no option to change the `font-style` or `font-weight`.

### Example output

You'll notice a number of oddities in this syntax highlighting. For instance, HTML tags _must_ be the same color as strings 😕 

image

### CODE REVIEW

**Comment by FredKSchott** [comment]
LGTM! Its nice to see that the style works well across both light and dark mode.

**Comment by sarah11918** [comment]
I'm super-impressed with all the customization! 

One note: ss someone who is VS-code challenged, is it normal for the units to be a different colour from the values (e.g. in 50px)? I'm new to this syntax highlighting thing, but that one is really bordering on "Why is that even a thing?" level of distracting to me. But, if it's what people expect, then give the people what they want!

Otherwise, I love the customization!

**Comment by FredKSchott** [comment]
Ah yea, I just checked my VSCode theme and it does the same (# and unit same color):

Screen Shot 2022-03-16 at 4 16 51 PM

also, `var` and `--var-name` being different colors feels a bit nicer to me.

**Comment by FredKSchott** [comment]
I love the idea of setting our own colors here, but as a far _how_ we color different tokens I'd like to follow the VSCode defaults as closely as possible. I'm sure they've thought through these things way more than we have, and we don't really gain much by disrupting how peoples eyes have been trained (aka not a hill we need to die on :)

@bholmesdev are you okay to update some of these rules to better match VSCode defaults.

File: public/index.css [hunk 1]
```diff
@@ -181,14 +181,11 @@ strong {
 
 /* Supporting Content */
 code {
-	font-family: var(--font-mono);
-	font-size: 0.85em;
-}
-
-code:not([class*='language']) {
```
Comment (FredKSchott) [inline public/index.css:188]: Are you sure about removing this? Just confirming since this previously styled inline code but not larger code blocks.

**Review by FredKSchott — APPROVED** [review]
LGTM with 1 thing to confirm

**Comment by bholmesdev** [comment]
> Looks great to me! I notice there's no Astro component in the most recent example. What colour are they? Are they the same as HTML elements, or different?

Feel free to [browse the deploy preview](https://deploy-preview-235--astro-docs-2.netlify.app/en/core-concepts/astro-components/) to verify! I assume you're wondering if `<Components />` have different styles from `<elements>`? Because yes, they will appear as a different color. This matches our existing syntax highlighting.

image

**Comment by bholmesdev** [comment]
> LGTM with 1 thing to confirm

Yep, these styles are getting reset! The problem with targeting inline code is this:
- Prism applies a `language-*` class directly on the `code` element
- Shiki applies an `astro-code` class on the outer `pre` element. The `code` element lacks any classes
☝️ This rules out using `:not(...)`. To avoid conflicts, we need to reset styles back to their defaults whenever we target inline `code`. @natemoo-re pointed out we can use `all: unset` for the simplest ruleset. See the [latest commit here](https://github.com/withastro/docs/pull/235/commits/1794905825ea136d512cf86b483c714559c20432)!

**Comment by FredKSchott** [comment]
LGTM!

**Pull request merged**